### PR TITLE
fix matrix multiplication interaction with HermOrSym and Diagonal

### PR DESCRIPTION
**Pull request by fredrikekre**
Fix https://github.com/JuliaLang/julia/pull/19228#issuecomment-310289197

### CODE REVIEW

File: base/linalg/diagonal.jl [hunk 1]
```diff
@@ -150,9 +150,9 @@ end
 (*)(D::Diagonal, B::AbstractTriangular) = A_mul_B!(D, copy(B))
 
 (*)(A::AbstractMatrix, D::Diagonal) =
-    scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag))), A, D.diag)
+    scale!(similar(A, promote_op(*, eltype(A), eltype(D.diag)), size(A)), A, D.diag)
```
Comment (fredrikekre) [inline base/linalg/diagonal.jl]: This fix is a bit dirty; currently `similar(A::Symmetric)` return a `Matrix` iff you supply both a new element type and size.
Ref: #19016 #15198 and #13731

Comment (fredrikekre) [inline base/linalg/diagonal.jl]: Seem like this is used elsewhere also, so I think this is OK.

Comment (tkelman) [inline base/linalg/diagonal.jl]: There are some ongoing open API issues between similar vs "copystructure" but for now I think it's okay

**Review by Sacha0 — APPROVED** [review]
lgtm! :)

**Comment by fredrikekre** [comment]
Thanks for reviewing/merging.

**Pull request merged**